contracts/cw-ics721-bridge/src/contract.rs

```diff
@@ -0,0 +1,451 @@
+    info: MessageInfo,
+    class_id: String,
+    token_id: String,
+    receiver: String,
+) -> Result<Response, ContractError> {
+    // This will error if the class_id does not exist so no need to check
+    let owner = get_owner(deps, class_id.clone(), token_id.clone())?;
+
+    // Check if we are the owner or the contract itself
+    if info.sender != env.contract.address && info.sender != owner.owner {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let msg = transfer(deps, class_id, token_id, receiver)?;
+    Ok(Response::new()
+        .add_attribute("action", "transfer")
+        .add_submessage(msg))
+}
+
+fn execute_burn(
+    deps: Deps,
+    env: Env,
+    info: MessageInfo,
+    class_id: String,
+    token_id: String,
+) -> Result<Response, ContractError> {
+    // This will error if the class_id does not exist so no need to check
+    let owner = get_owner(deps, class_id.clone(), token_id.clone())?;
+
+    // Check if we are the owner or the contract itself
+    if info.sender != env.contract.address && info.sender != owner.owner {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let msg = burn(deps, class_id, token_id)?;
+    Ok(Response::new()
+        .add_attribute("action", "burn")
+        .add_submessage(msg))
+}
+
+fn execute_mint(
+    deps: Deps,
+    env: Env,
+    info: MessageInfo,
+    class_id: String,
+    token_ids: Vec<String>,
+    token_uris: Vec<String>,
+    receiver: String,
+) -> Result<Response, ContractError> {
+    if info.sender != env.contract.address {
+        return Err(ContractError::Unauthorized {});
+    }
+    if token_ids.len() != token_uris.len() {
+        return Err(ContractError::ImbalancedTokenInfo {});
+    }
+    let receiver = deps.api.addr_validate(&receiver)?;
+    let cw721_addr = CLASS_ID_TO_NFT_CONTRACT.load(deps.storage, class_id)?;
+
+    let mint_messages = token_ids
+        .into_iter()
+        // Can zip without worrying about dropping data as we assert
+        // that lengths are the same above.
+        .zip(token_uris.into_iter())
+        .map(|(token_id, token_uri)| -> StdResult<WasmMsg> {
+            let msg = cw721_base::msg::ExecuteMsg::Mint(cw721_base::MintMsg::<Empty> {
+                token_id,
+                token_uri: Some(token_uri),
+                owner: receiver.to_string(),
+                extension: Empty::default(),
+            });
+            Ok(WasmMsg::Execute {
+                contract_addr: cw721_addr.to_string(),
+                msg: to_binary(&msg)?,
+                funds: vec![],
+            })
+        })
+        .collect::<StdResult<Vec<_>>>()?;
+
+    Ok(Response::default()
+        .add_attribute("method", "execute_mint")
+        .add_messages(mint_messages))
+}
+
+#[allow(clippy::too_many_arguments)]
+fn execute_do_instantiate_and_mint(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    class_id: String,
+    class_uri: Option<String>,
+    token_ids: Vec<String>,
```

Comment:
Was this always meant to be a batch method? Before I think it was a single token id ?

Reply:
It needs to be a batch method because it deals with minting new tokens upon receiving a ICS 721 NFT. Those packets can send multiple NFTs at once (`token_ids` is a list), so we must be able to mint multiple at once.

Reply:
yepp, the original spec does in fact say token_ids